README.md
# els - Extended LS
## History
This is a fork of the work found at:

- els-1.54a1.tar.gz from http://els-software.org

I have modified it for my own person use.

*No claims are made to the reliability or safety of those changes.*

Extended LS is a very extended version of 'ls' by the original author.  I have only made a few minor tweaks, such as adding color and glyph support.

README-ORIG is the original README from els-software.org.
All of the other text files and documentation are unmodified from the original source.

## Compiling
This has compiled on OS X, Big Sur, not tested on other platforms.

- make config
- make all

binaries will be in:  Build/els/bin

## Features
- add commas to file sizes
- colorize file size column
- colorize date/time by age
- add glyphs to show file types

## Using
These are zsh functions I use to call it.

Note: You will need to remove the "| mc" or also install it.
mc turns "line" output into multi-columns

- l  () { els +T^NY-M-DT +G~Atp~ugsmNL $@ }
![l]<img/l.png>

- ll () { els +T^NY-M-DT +G~At~smN $@ | mc }
![ll]<img/ll.png>

- ls () { els +G~t~N $@ | mc }
![ls]<img/ls.png>

- lt () { els +T^NY-M-DT +G~At~smN -rt $@ | mc }
![lt]<img/lt.png>


NEWS
January 9, 2009:


EXCITING NEWS: ELS is now on the WEB after years of PROCRASTINATION!

Please visit "els-software.org" for all your ELS needs and late-breaking
news.


MAILING LIST:

Please send email to "requestXXX@els-software.org (remove XXX)" requesting
to be placed on the "ELS ANNOUNCEMENT" mailing list if you'd like to receive
announcements regarding major releases and/or significant changes to ELS.
This mailing list will be very low bandwidth and you should receive no more
than 2 or 3 emails per millennium.


SIGNIFICANT CHANGES AS OF VERSION 1.50:

5) Added locale awareness to "els" (regular listings only) and "edate"

4) ACL support added for Linux2.6+

3) Makefile now uses output from "uname -s" as the OS target as well
   as significant restructuring and cleanup.

2) Added config program for automatically determining formats and sizes
   of time_t, st_ino, st_size, st_nlink, st_blocks at compile time.

1) Fixed checksums of files larger than 2^31 by using open64().


SIGNIFICANT CHANGES AS OF VERSION 1.49:

4) Quotaling (+Q option) now supports 32-bit UID and GIDs as of
   release 1.49h1!

3) Replaced sorting algorithms with quick-sort (extremely large
   directories now list up to 30 times faster!).

2) 1.49a2 fixes "-b" bug for characters >= 0x80

1) New option +c for displaying environment as part of the listing


SIGNIFICANT CHANGES AS OF VERSION 1.48:

7) Checksum algorithm option added as of 1.48f1.  E.g.:
	els +C=bsd +z +GCSN -RA		# similar to /usr/ucb/sum
	els +C=sysv +5l +z +GCSN -RA	# similar to /usr/bin/sum
	els +C=posix +z +GCsN -RA	# similar to /usr/bin/cksum
	els +C=crc32 +z +GCsN -RA	# crc32 algorithm

6) Cygwin is supported as of 1.48e1 (tested using cygwin 1.3.22 and 1.5.20)

5) Added floating point recognition and 'KkMmGgTt' (i.e. Kilo/Mega/Giga/
   Tera) modifiers to F'Q{...}' expressions.  For example, to list
   all files greater than 1.1M that were modified over half a year ago:
	els +F'Q{s>1.1M & m>.5Y}' -laR

4) New option +Q: Display Quotals (quota+total) for indicated hierarchy or
   files.  The quotal option will list the total size and number of files
   owned by each user/UID and group/GID.

3) New option +o: Specify file for output, e.g.:  els -laR +o=output.dat

2) The syntax of +t has evolved.  '+t=M' means to traverse mount-points,
   '+t=L' means to traverse expanded symbolic-links, '+t=LM' means to
   traverse both mount-points and expanded symbolic-links.  Note that
   '+t=L' is  effective only if '-L' is also specified, as '-L' is what
   enables expansion of symbolic-links.

1) ELS now returns a non-zero value if *any* errors occur during listing.


SIGNIFICANT CHANGES AS OF VERSION 1.47:

8) A Watch function (+W) has been added added as of release 1.47k1 allowing
   you to watch ELS's progress when processing large hierarchies containing
   numerous files (this option works best in conjunction with +F filtering
   when output is sparse and processing time is lengthy or when redirecting
   stdout to a file).  +W also interactively intercepts SIGQUIT (i.e. ^\)
   and will display the directory name currently being processed.

7) Version 1.47e1+ fixes a +F filtering problem under Linux.
   Version 1.47k1+ fixes a Segmentation fault problem under OSF1.

6) The -F flag was added to chdate and is now required if changing multiple
   files as of 1.47h1; if changing a single file then the -F flag is
   optional.  Moreover, chdate no longer requires quoting date arguments
   since the -F flag now helps distinguish between dates and files.  E.g.,
   the following are now equivalent:
	chdate Apr 1 12:34 file1
	chdate "Apr 1 12:34" file1
   Here the -F flag must be specified since multiple files are given:
	chdate -a Apr 1 2000 12:34 -F file1 file2 file3

5) The unusual filter function for ELS now supports specifying sparse
   files (i.e. files partially populated with data) as of 1.47g1.  For
   example, to list all sparse files under the mounted file systems /,
   /usr, and /var:
	els +F'U{s}' -laR / /usr /var

4) The option +Z=zonename was added as of release 1.47c2 allowing the TZ
   (i.e. timezone) setting to be changed during execution of els, chdate,
   and edate.

3) +Td (dos-style) and +Tw (windows-style) date formats added as
   of release 1.47b.

2) The meaning of +t has changed as of release 1.47b.  Specifying +t now
   causes file-systems to be traversed whereas before it inhibited traversal.
   Unless you want to potentially list multiple file-systems on your local
   host or across your network, you probably don't want to use the +t option.
   Changing the meaning of +t was an agonizing decision but was necessary as
   future releases will allow +t=<type> specifying the types of directories
   and/or file-systems to traverse.

1) The listing of large files (i.e. filesize > 2**31) under SunOS5.6+ and
   HPUX10+ is now supported as of release 1.47b.


UNIX NOW MORE THAN 10^9 SECONDS OLD:

Unix is now more than one-billion seconds old!  Execute the following
"edate" command to discover when this milestone was reached:
	edate -C 1000000000 +Z=GMT0


FTP/HTTP ACCESS:

There are several alternate sites that are hosting copies of ELS, but
keep in mind that their copies will probably not be as current as my
copy.  To download my most recent version please reference:

Most current version:
	http://els-software.org

Mirrored versions:
	http://www.sfr-fresh.com/els

Additionally, the HPUX users group carrires a pre-packaged version of ELS
specifically tailored for HPUX (but which is out-of-date):

	http://hpux.ee.ualberta.ca/hppd/hpux/Users/els-1.48a/
	(NB: this pre-packaged version is specific to HPUX and may
	present difficulties if used on non-HPUX systems)

Markb.


cksum.c
/******************************************************************************
  cksum.c -- checksum functions

  Author: Mark Baranowski
  Email:  requestXXX@els-software.org (remove XXX)
  Download: http://els-software.org

  Last significant change: September 8, 2015
  
  These functions are provided "as is" in the hopes that they might serve some
  higher purpose.  If you want these functions to serve some lower purpose,
  then that too is all right.  So as to keep these hopes alive you may
  freely distribute these functions so long as this header remains intact.
  You may also freely distribute modified versions of these functions so long
  as you indicate that such versions are modified and so long as you
  provide access to the unmodified original copy.

  NOTE: the POSIX and CRC32 algorithms contained here-in do NOT rely on
  pre-computed CRC tables (it took me way too long searching the 'Net to
  find versions NOT based on pre-computed CRC tables, and when I did find
  something close to what I wanted I still had to modify them so as to do
  run-time initialization!).  I do not like pre-computed CRC tables because:
  a long list of binary numbers isn't meaningful to a human, it takes
  very little time to initialize these tables at run-time, and generating
  them at run-time avoids portability issues between big/little-endian and
  32/64-bit machines.

  ACKNOWLEDGEMENTS:
  1) Mark Baranowski is the author of cksumBsd() and cksumSysv().  I
     hereby give myself (and everyone else) permission to use my functions.
  2) The POSIX checksum functions fill_r(), remainder(),
     and cksumPosix() were adapted from GNU's coreutils-5.0.tar.gz.
     Please see GNU copyright header below pertaining to these functions.
  3) THE ANSI/CRC32 checksum functions cksumCrc32Init() and cksumCrc32()
     were adapted from ZLIB's crc32.c funcion.  Please see ZLIB copyright
     header below pertaining to these functions.

  Note: The most recent version of these functions may be obtained from
        http://els-software.org

  ****************************************************************************/

#include "sysdefs.h"

#include <stdio.h>
#include <errno.h>
#if defined(HAVE_O_NOATIME) && defined(LINUX)
/* __USE_GNU allows O_NOATIME */
/* TBD: Does __USE_GNU introduce other gremlins? */
# define __USE_GNU
# include <fcntl.h>
# undef  __USE_GNU
#else
# include <fcntl.h>
#endif
#include <sys/types.h>
#include <unistd.h>

#include "defs.h"
#include "auxil.h"
#include "cksum.h"
#include "elsVars.h"

/* NB: sizeof(long) does not always == 4, e.g. on OSF1 sizeof(long) == 8 */
#if ELS_SIZEOF_st_size <= ELS_SIZEOF_long
#define MAX_ST_SIZE  MAXPOS_LONG
#else
#define MAX_ST_SIZE  MAXPOS_LLONG
#endif
#define MAX_READ     0x2000


/* The "read size" must carefully match the OS's limitations, because some
   OSes (such as SunOS4.x) have problems if attempting to read more data
   than the size of the file when the file's size is between 0x7fffe000 and
   0x7fffffff (even if using getc() this is still a problem because stdio
   reads in blocks of 0x2000 and since the final block exceeds the maximum
   allowed file size, errno = EINVAL gets set!).  The function read_size()
   prevents this because it limits the final block "read size" to 0x1fff. */

Local int read_size(ELS_st_size offset)
{
  /* NB: "ELS_st_size" is assumed to be unsigned */
  if (offset >= MAX_ST_SIZE)
    return(0);  /* Already past MAX_ST_SIZE */
  else if (offset + MAX_READ > MAX_ST_SIZE)
    return((int)(MAX_ST_SIZE - offset));  /* Don't exceed MAX_ST_SIZE */
  else
    return(MAX_READ);  /* OK to read a full buffer */
}


/* Convert cksumType to algorithm name: */
char *cksumTypeToName(cksumType type)
{
  char *name;
  if (type == CKSUM_BSD)
    name = "bsd";
  else if (type == CKSUM_SYSV)
    name = "sysv";
  else if (type == CKSUM_POSIX)
    name = "posix";
  else if (type == CKSUM_CRC32)
    name = "crc32";
  else
    name = "unknown";
  return(name);
}


/* Both /usr/ucb/sum and /usr/bin/sum display 0 if error occurs.
   crc32 displays 0 if error occurs just because.
   SunOS /usr/bin/cksum displays 4294967295 if cksum of directory, while
   Linux's /usr/bin/cksum prints error message and no value in such cases.
   Prior to VersionLevel 152 els displayed 4294967295 (unless +z in effect),
   but since zero byte files also return 4294967295 it makes better sense
   to display 0 like everyone else. */

Uint32 cksumErrorCode(cksumType type)
{
  extern int VersionLevel;  /* From els.h */
  if (type == CKSUM_POSIX && VersionLevel <= 151)
    /* Beware of x64_64 Linux where sizeof(Ulong) == 8: */
    return (Uint32)0xffffffff;
  else
    return 0;
}


/* Perform checksum of file using requested algorithm: */
Uint32 cksumFile(char *filename, cksumType type, ELS_st_size *nread)
{
  int fd;
  Uint32 checksum = 0;
  if (nread != NULL) *nread = 0;

#if defined(HAVE_O_NOATIME)
  /* TBD: Does O_NOATIME perturb non-Linux FSes? */
  {
    int save_errno = errno;
    static Boole give_warning = TRUE;
    fd = open(filename, O_RDONLY|O_NOATIME);
    if (fd < 0)
    {
      /* Discard errno from previous attempt and re-try without O_NOATIME: */
      errno = save_errno;
      fd = open(filename, O_RDONLY);
      if (fd >= 0 && verboseLevel >= 1 && give_warning)
      {
	fprintf(stderr, "\
%s: FYI: Unable to preserve access time: %s\n\
", Progname, filename);
	if (verboseLevel == 1)
	{
	  fprintf(stderr, "\
%s: Subsequent access time warnings suppressed\n\
", Progname);
	  give_warning = FALSE;
	}
      }
    }
  }
#else
  fd = open(filename, O_RDONLY);
#endif
  if (fd >= 0)
  {
    if (type == CKSUM_BSD)
      checksum = cksumBsd(fd, nread);    /* Same algorithm as /usr/ucb/sum */
    else if (type == CKSUM_SYSV)
      checksum = cksumSysv(fd, nread);   /* Same algorithm as /usr/bin/sum */
    else if (type == CKSUM_POSIX)
      checksum = cksumPosix(fd, nread);  /* Same algorithm as /usr/bin/cksum */
    else if (type == CKSUM_CRC32)
      checksum = cksumCrc32(fd, nread);  /* CRC32 algorithm */
    close(fd);
  }
  
  return(checksum);
}


/* PERFORM 16-BIT BSD CHECKSUM (same algorithm as /usr/bin/sum under SunOS4,
   /usr/ucb/sum under SunOS5/Solaris, and '/usr/bin/sum -r' under Linux): */
Uint32 cksumBsd(int fd, ELS_st_size *nread)
{
  register Uint32 checksum = 0;
  Uchar buf[MAX_READ];
  ELS_st_size offset = 0;
  int n;
  int save_errno = errno;
  errno = 0;

  while ((n = read(fd, buf, read_size(offset))) > 0)
  {
    register Uchar *cp = buf;
    offset += n;
    while (n-- > 0)
    {
      Boole rotate = (checksum & 1) != 0;
      checksum >>= 1;
      if (rotate) checksum |= 0x8000;
      checksum += *cp++;
      checksum &= 0xffff;
    }
  }

  if (errno != 0)
    checksum = cksumErrorCode(CKSUM_BSD);  /* Keep current errno */
  else
    errno = save_errno;  /* Restore previous errno */

  if (nread != NULL) *nread = offset;
  return(checksum);
}


/* PERFORM 16-BIT SYSV CHECKSUM (same algorithm as /usr/5bin/sum under SunOS4,
   /usr/bin/sum under SunOS5/Solaris, and '/usr/bin/sum -s' under Linux): */
Uint32 cksumSysv(int fd, ELS_st_size *nread)
{
  register Uint32 checksum = 0;
  Uchar buf[MAX_READ];
  ELS_st_size offset = 0;
  int n;
  int save_errno = errno;
  errno = 0;

  while ((n = read(fd, buf, read_size(offset))) > 0)
  {
    register Uchar *cp = buf;
    offset += n;
    while (n-- > 0)
    {
      checksum += *cp++;
    }
  }

  if (sizeof(checksum) > 4) checksum &= 0xffffffff;
  checksum = (checksum & 0xffff) + (checksum >> 16);
  checksum = (checksum & 0xffff) + (checksum >> 16);

  if (errno != 0)
    checksum = cksumErrorCode(CKSUM_SYSV);  /* Keep current errno */
  else
    errno = save_errno;  /* Restore previous errno */

  if (nread != NULL) *nread = offset;
  return(checksum);
}


/*****************************************************************************/
/* PERFORM 32-BIT POSIX CHECKSUM (same algorithm as /usr/bin/cksum under
   both SunOS5/Solaris and Linux.  See also Posix1003.2/D11.2): */

/* cksum -- calculate and print POSIX checksums and sizes of files
   Copyright (C) 92, 1995-2003 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Written by Q. Frank Xia, qx@math.columbia.edu.
   Cosmetic changes and reorganization by David MacKenzie, djm@gnu.ai.mit.edu.
   
   As Bruce Evans pointed out to me, the crctab in the sample C code
   in 4.9.10 Rationale of P1003.2/D11.2 is represented in reversed order.
   Namely, 0x01 is represented as 0x80, 0x02 is represented as 0x40, etc.
   The generating polynomial is crctab[0x80]=0xedb88320 instead of
   crctab[1]=0x04C11DB7.  But the code works only for a non-reverse order
   crctab.  Therefore, the sample implementation is wrong.
   
   This software is compatible with neither the System V nor the BSD
   `sum' program.  It is supposed to conform to P1003.2/D11.2,
   except foreign language interface (4.9.5.3 of P1003.2/D11.2) support.
   Any inconsistency with the standard except 4.9.5.3 is a bug.  */

/* Adapted to ELS by Mark Baranowski, requestXXX@els-software.org (remove XXX)
   THIS VERSION HAS BEEN MODIFIED FROM THE ORIGINAL.  The original cksum.c
   can be found in coreutils-5.0.tar.gz (available from http://www.gnu.org).
   */

# define BIT(x)	((Uint32) 1 << (x))
# define SBIT	BIT(31)

/* The generating polynomial is:

         32  26  23  22  16  12  11  10  8   7   5   4   2   1   0
   G(X)=X + X + X + X + X + X + X + X + X + X + X + X + X + X + X

   The i bit in POSIX_GEN is set if X^i is a summand of G(X) except X^32. */

#define POSIX_GEN (BIT(26) | BIT(23) | BIT(22) | BIT(16) | BIT(12) | \
		   BIT(11) | BIT(10) | BIT(8) | BIT(7) | BIT(5) | \
		   BIT(4) | BIT(2) | BIT(1) | BIT(0))

Local Uint32 *cksumPosix_crctab = NULL;
Local Uint32 r[8];


Local void cksumPosix_fill_r(Uint32 GEN)
{
  int i;

  r[0] = GEN;
  for (i = 1; i < 8; i++)
    r[i] = (r[i - 1] << 1) ^ ((r[i - 1] & SBIT) ? GEN : 0);
  return;
}


Local Uint32 cksumPosix_remainder(int m)
{
  Uint32 rem = 0;
  int i;

  for (i = 0; i < 8; i++)
    if (BIT(i) & m)
      rem ^= r[i];

  return(rem & 0xFFFFFFFF);	/* Make it run on 64-bit machine.  */
}


Local void cksumPosixInit(Uint32 GEN)
{
  if (cksumPosix_crctab == NULL)
  {
    int i;
    cksumPosix_crctab = memAlloc(256 * sizeof(Uint32));

    cksumPosix_fill_r(GEN);
    cksumPosix_crctab[0] = 0;
    for (i = 1; i < 256; i++)
      cksumPosix_crctab[i] = cksumPosix_remainder(i);
  }
  return;
}


Uint32 cksumPosix(int fd, ELS_st_size *nread)
{
  register Uint32 crc = 0;
  Uchar buf[MAX_READ];
  ELS_st_size offset = 0;
  int n;
  int save_errno = errno;
  errno = 0;

  if (cksumPosix_crctab == NULL)
    cksumPosixInit(POSIX_GEN);

  while ((n = read(fd, buf, read_size(offset))) > 0)
  {
    register Uchar *cp = buf;
    offset += n;
    while (n-- > 0)
    {
      crc = (crc << 8) ^ cksumPosix_crctab[((crc >> 24) ^ *cp++) & 0xFF];
    }
  }

  if (nread != NULL) *nread = offset;
  
  for (; offset; offset >>= 8)
    crc = (crc << 8) ^ cksumPosix_crctab[((crc >> 24) ^ offset) & 0xFF];

  crc = ~crc & 0xFFFFFFFF;

  if (errno != 0)
    crc = cksumErrorCode(CKSUM_POSIX);  /* Keep current errno */
  else
    errno = save_errno;  /* Restore previous errno */

  return(crc);
}


/*****************************************************************************/
/* PERFORM 32-BIT CRC32 CHECKSUM (same algorithm as used by ZLIB, et al.
   See also ANSI X3.66): */

/* crc32.c -- compute the CRC-32 of a data stream
 * Copyright (C) 1995-2002 Mark Adler
 * For conditions of distribution and use, see copyright notice in zlib.h 
 */

/* zlib.h -- interface of the 'zlib' general purpose compression library
  version 1.1.4, March 11th, 2002

  Copyright (C) 1995-2002 Jean-loup Gailly and Mark Adler

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  Jean-loup Gailly        Mark Adler
  jloup@gzip.org          madler@alumni.caltech.edu
*/

/*
   Generate a table for a byte-wise 32-bit CRC calculation on the polynomial:
   x^32+x^26+x^23+x^22+x^16+x^12+x^11+x^10+x^8+x^7+x^5+x^4+x^2+x+1.
   
   Polynomials over GF(2) are represented in binary, one bit per coefficient,
   with the lowest powers in the most significant bit.  Then adding polynomials
   is just exclusive-or, and multiplying a polynomial by x is a right shift by
   one.  If we call the above polynomial p, and represent a byte as the
   polynomial q, also with the lowest power in the most significant bit (so the
   byte 0xb1 is the polynomial x^7+x^3+x+1), then the CRC is (q*x^32) mod p,
   where a mod b means the remainder after dividing a by b.
   
   This calculation is done using the shift-register method of multiplying and
   taking the remainder.  The register is initialized to zero, and for each
   incoming bit, x^32 is added mod p to the register if the bit is a one (where
   x^32 mod p is p+x^32 = x^26+...+1), and the register is multiplied mod p by
   x (which is shifting right by one and adding x^32 mod p if the bit shifted
   out is a one).  We start with the highest power (least significant bit) of
   q and repeat for all eight bits of q.
   
   The table is simply the CRC of all possible eight bit values.  This is all
   the information needed to generate CRC's on data a byte at a time for all
   combinations of CRC register values and incoming bytes.
   */

/* Adapted to ELS by Mark Baranowski, requestXXX@els-software.org (remove XXX)
   THIS VERSION HAS BEEN MODIFIED FROM THE ORIGINAL.  The original crc32.c
   can be obtained from http://www.zlib.org.
   */

Local Uint32 *cksumCrc32_crctab = NULL;


Local void cksumCrc32Init(void)
{
  Uint32 c;
  int n, k;
  Uint32 poly;            /* polynomial exclusive-or pattern */
  /* terms of polynomial defining this crc (except x^32): */
  static const char p[] = {0,1,2,4,5,7,8,10,11,12,16,22,23,26};
  
  if (cksumCrc32_crctab == NULL)
  {
    cksumCrc32_crctab = memAlloc(256 * sizeof(Uint32));
    
    /* make exclusive-or pattern from polynomial (0xedb88320L) */
    poly = 0L;
    for (n = 0; n < sizeof(p); n++)
      poly |= 1L << (31 - p[n]);
    
    for (n = 0; n < 256; n++)
    {
      c = (Uint32)n;
      for (k = 0; k < 8; k++)
	c = c & 1 ? poly ^ (c >> 1) : c >> 1;
      cksumCrc32_crctab[n] = c;
    }
  }
  return;
}


Uint32 cksumCrc32(int fd, ELS_st_size *nread)
{
  register Uint32 crc = 0;
  Uchar buf[MAX_READ];
  ELS_st_size offset = 0;
  int n;
  int save_errno = errno;
  errno = 0;

  if (cksumCrc32_crctab == NULL)
    cksumCrc32Init();

  crc = crc ^ 0xffffffffL;

  while ((n = read(fd, buf, read_size(offset))) > 0)
  {
    register Uchar *cp = buf;
    offset += n;
    while (n-- > 0)
    {
      crc = cksumCrc32_crctab[((int)crc ^ *cp++) & 0xff] ^ (crc >> 8);
    }
  }

  if (errno != 0)
    crc = cksumErrorCode(CKSUM_CRC32);  /* Keep current errno */
  else
    errno = save_errno;  /* Restore previous errno */

  if (nread != NULL) *nread = offset;
  return(crc ^ 0xffffffffL);
}
